### PR TITLE
SqueezeSeg reprodroduction

### PR DESCRIPTION
**Pull request by jdaaph**
Reproduce code from SqueezeSeg (ICRA 18')

### CODE REVIEW

File: cli.py [hunk 1]
```diff
@@ -1,7 +1,7 @@
 import click
 
 import ml
-from ml.base import MLConfig
+from ml.base import GlobalConfig
 from ml.base.estimator import MLEstimator
 from ml.squeezeseg.cli import npy2tfrecord
```
Comment (jdaaph) [inline cli.py]: this import is very ugly @FXuZ do you have a better solution to make sure the class definition for specific io / trainers etc run before cli's logic is called?

File: ml/base/estimator.py [hunk 2]
```diff
@@ -52,31 +41,34 @@ def _gen_estimator(self):
         )
         return estimator
 
-    def _gen_model_fn(self):
+    def _gen_model_fn(self, gpu_id=2):
 
         def model_fn(features, labels, mode, params):
```
Comment (FXuZ) [inline ml/base/estimator.py]: Why is a `model_fn` defined in the prototype class? I assume that the abstract classes are not supposed to have specific implementations.

File: ml/base/estimator.py [hunk 3]
```diff
@@ -52,31 +41,34 @@ def _gen_estimator(self):
...
-            # Construct graph.
-            self.output = self.graph.add_forward_pass()
+            # TODO(jdaaph): Add cpu and gpu flag.
+            with tf.device('/gpu:{}'.format(gpu_id)):
```
Comment (FXuZ) [inline ml/base/estimator.py]: Is there a more elegant way of registering variables to some certain device / namespace of tf?

File: ml/base/config.py [hunk 4]
```diff
@@ -46,3 +37,36 @@ def __init__(self, global_config, graph, trainer, predictor, io, metric):
...
+        component_config_lst = [global_config]
+
+        for component in ['graph', 'trainer', 'predictor', 'io', 'metric']:
+            config = munch.munchify(getattr(global_config, component))
+            component_config_lst.append(config)
+
+        return GlobalConfig(*component_config_lst)
+
+
+class CompConfig(MLConfig):
```
Comment (FXuZ) [inline ml/base/config.py]: Can we add a doc string to this unimplemented class?

File: ml/base/estimator.py [hunk 5]
```diff
@@ -52,31 +41,31 @@ def _gen_estimator(self):
         )
         return estimator
 
-    def _gen_model_fn(self):
+    def _gen_model_fn(self, gpu_id=2):
```
[the remark this answers is not in the document]
Comment (FXuZ) [reply ml/base/estimator.py]: I guess that it's time for us to add a few more global fields in the config file. And we can let modules that change the computational graph deal with the device and tf namespace stuff.

File: cli.py (hunk 1 above)
Comment (jdaaph) [reply cli.py]: procrastinated to issue #27

File: ml/base/estimator.py (hunk 2 above)
Comment (jdaaph) [reply ml/base/estimator.py]: My plan is to let custom components implement the used methods in _gen_model_fn.

File: ml/base/estimator.py (hunk 3 above)
Comment (jdaaph) [reply ml/base/estimator.py]: I put that as an issue #28

File: ml/base/config.py (hunk 4 above)
Comment (jdaaph) [reply ml/base/config.py]: I just deleted it

**Pull request merged**